### DISCUSSION TITLE
Possibility of manipulating local context in batch loaders

### DISCUSSION THREAD

**Discussion by marceloverdijk** [graphql-java/graphql-java#2193] [Q&A] [answered]
First of all I think the concept of Passing Data and Local Context as described in [this ](https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/) blog post is worth to be added to reference guide.

I found out about the local context in api docs, but was pointed to the blog post by someone else and which was very useful.

While trying the local context it worked quite well, until the point `BatchLoader` came into the picture which fetched an object from which I also wanted to push some context down to the child fetchers. Unfortunately this is not possible - or at least I did not found a way to do it.

Let me also try to explain with a fictive setup. Imagine a schema like:

```
type Person {
    id: ID!
    name: String!
    countryOfBirth: Country
    nationality: Country!
}

type Country {
    id: ID!
    code: String!
    name: String!
    continent: Continent!
}

type Continent {
    id: ID!
    name: String!
}
```

When fetching the `Person` the `countryOfBirth` and `nationality` will not be fetched automatically, but in the `PesonDataFetcher` I would store both the `countryOfBirthCountryId` and `nationalityCountryId`.

In the `countryOfBirth` and `nationality` data fetchers I'm using a `BatchLoader` to fetch all needed countries at once, but without the `Continent` this time. So here I would like to store the `continentId` in the local context, so that in the `continent` data fetcher I would know which continent to fetch.

Is this already possible or would there be a way to do this?

**Comment by bbakerman** [comment, marked as answer]
DataLoader as a library is not graphql specific - so it does not know about graphql-java things like localContext

There is this in data loader

```
    public CompletableFuture<V> load(K key, Object keyContext) {
```

That is when you load an entity you can pass in a context.

The batch loader is then

```
public interface BatchLoaderWithContext<K, V> {
    /**
     * Called to batch load the provided keys and return a promise to a list of values.  This default
     * version can be given an environment object to that maybe be useful during the call.  A typical use case
     * is passing in security credentials or database details for example.
     *
     * @param keys        the collection of keys to load
     * @param environment an environment object that can help with the call
     *
     * @return a promise of the values for those keys
     */
    CompletionStage<List<V>> load(List<K> keys, BatchLoaderEnvironment environment);
}
```

and

```
public class BatchLoaderEnvironment {

    private final Object context;
    private final Map<Object, Object> keyContexts;
    private final List<Object> keyContextsList;

```

So on INPUT you can pass down context objects.  

HOWEVER DataLoader does not have the concept of a `DataFetcherResult` where you can return a data object AND other data like errors and context (this might be a good idea muse on it) - not natively

But if you return `DataFetcherResult`objects  and these make it back into graphql land then graphql-java will interpret them and pull them apart and set context for child fetchers.

So you batch loader would be

```
new BatchLoader<YourKeyType, DataFetcherResult<YourValueType>>()
```

Therefore when you batch load Country you can store away a `continentId` per object returned in the list of results

Does this help.

**Reply by marceloverdijk** [reply]
Thx @bbakerman 

For a bit of context I was playing with the Netflix DGS framework and now used another pattern (as discussed [here](https://github.com/Netflix/dgs-framework/discussions/21)).
Meaning to extend generated types classes to contain the additional needed information.

I'm going to try your suggested approach with the local context in combination with the data loaders to see which pattern I like the most.

**Comment by bbakerman** [comment]
I might make this a discussion rather than an issue
